### DISCUSSION TITLE
Custom DB table as main source for bucket

### DISCUSSION THREAD

**Discussion by setofaces** [vladimir-bukhtoyarov/bucket4j#382] [Q&A] [answered]
Hello there, I am bit confused what is the best strategy to use your library in my case or maybe don't use it at all:
There is  db table with montly limit value and current request counter for each user, naively I saw firstly things may be done this way:  request is intercepted, appropriate bucket of user is loaded if exists or is created with capacity from a db table (montly limit - current counter). The main drawback here is necessity of maintating consistency between buckets and that table probably via BucketListener. But there is also postgresql integration already as I see so probably things may be easier. Does that stuff sounds like bullshit or overkill? 
I can'fully rely on cache because of risk losing data after application shutdown, but at the same time don't want to use some additional persistence storage like redis except my main postgre database. Should I rely on library built-in postgresql integration somehow, maybe there is way to make it work with my custom table, because default one doesn't seem to be fully suitable for my case? For instance, as key I want to have text, not bigint Any answers to my vebiage are very appreciated!

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
@setofaces 

You can configure table name and name for fields via `BucketTableSettings.customSettings`. Also, it is not documented yet, but ability to have custom type for primary key was added in the scope of `8.2` https://bucket4j.com/8.2.0/release-notes.html

It is example of table and primary key type cusomization:
```java
    // specify required table name and column names
    BucketTableSettings tableSettings = BucketTableSettings.customSettings("some_table", "some_id", "some_state");

    SQLProxyConfiguration<String> configuration = SQLProxyConfigurationBuilder.builder()
                .withTableSettings(tableSettings)
                .withPrimaryKeyMapper(PrimaryKeyMapper.STRING) // specify type of primary key
                .build(dataSource);
        ProxyManager<String> proxyManager = new PostgreSQLadvisoryLockBasedProxyManager<>(configuration);
```

```java

```

**Reply by setofaces** [reply]
thank you man, sorry for stupid question

**Reply by raghu298** [reply]
@vladimir-bukhtoyarov getting below error once i used above code and made primaryKey as String:

Internal Server Error: io.github.bucket4j.BucketExceptions$BucketExecutionException: org.postgresql.util.PSQLException: ERROR: function pg_advisory_xact_lock(character varying) does not exist\n  Hint: No function matches the given name and argument types. You might need to add explicit type casts.\n  Position: 8

**Reply by raghu298** [reply]
@setofaces How you resolved the primarykey as String issue?

**Reply by setofaces** [reply]
the most essential solution was to choose select for update implementation instead of one based on advisory locks

**Reply by setofaces** [reply]
advisory locks can be used only with one or two numbers at he moment you better read some information about it

**Reply by raghu298** [reply]
@setofaces Could you please give sample code for select for update?

**Reply by setofaces** [reply]
just use [PostgreSQLSelectForUpdateBasedProxyManager](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-postgresql/src/main/java/io/github/bucket4j/postgresql/PostgreSQLSelectForUpdateBasedProxyManager.java) https://github.com/bucket4j/bucket4j/tree/master/bucket4j-postgresql/src/main/java/io/github/bucket4j/postgresql   in your bucket4j configuration

**Reply by raghu298** [reply]
@setofaces  thanks for saving my time.
i have one more query i have requirements to ratelimit based on request per minute(RPM) or on Tokens(open ai token) consumed by client.
openaiTokensConsumed will be sum up each time for the client and will have ratelimit on reaching the limit.
RPM able to achieve but how to do ratelimit on other parameters like openaiTokensConsumed?

**Reply by setofaces** [reply]
I dont fully get your requirements, you probably need 2 rate-limitters, one just to restrict aggressive usage of your service (RPM) and another to increment/decrement tokens for each user? Then it should  not be combined in my opinion and there is no reason in underlying persistence storage for RPM.

**Reply by setofaces** [reply]
Spring-wise I would suggest you doing 2 HandlerInterceptors

**Reply by raghu298** [reply]
@setofaces May be i confused think like you have column tokenConsumed which will be updated each time.
Now if for any user that tokenConsumed reaches above 2000 per minutes then restrict the user.
I am into micronaut and have to use Filters, interceptors not working in micronaut which works in spring.
But my question was how to ratelimit on the tokenconumed column?

**Reply by setofaces** [reply]
I am not a expert here, but do you really need database here? If your application is the only jvm instance running I dont see reasons to use database

**Reply by raghu298** [reply]
@setofaces Its distributed environment can run in multiple pods and can have Metrics too for knowing quota consumption

**Reply by setofaces** [reply]
Okay, then just refill tokens every minute and consume each request nothing compex here

**Reply by raghu298** [reply]
@setofaces May be not able to explain you i have to do 2 rateLimit.
1. RPM
2. on another ratelimit column called chocolatesConsumed(which will be updated from rest call eachtime), once chocolatesConsumed reaches 2000 then restrict the user.

**Reply by raghu298** [reply]
@setofaces i think its feasible but need to write 2 filters

**Reply by raghu298** [reply]
@setofaces  have you monitored the usage of bucket or the metrics?
